overlord/devicestate/firstboot.go

```diff
@@ -109,27 +91,23 @@ func populateStateFromSeedImpl(st *state.State, tm timings.Measurer) ([]*state.T
 		return nil, err
 	}
 
-	seedYamlFile := filepath.Join(dirs.SnapSeedDir, "seed.yaml")
-	if release.OnClassic && !osutil.FileExists(seedYamlFile) {
+	err = deviceSeed.LoadMeta(tm)
+	if release.OnClassic && err == seed.ErrNoMeta {
 		// on classic it is ok to not seed any snaps
 		return trivialSeeding(st, markSeeded), nil
 	}
-
-	seedMeta, err := seed.ReadYaml(seedYamlFile)
 	if err != nil {
 		return nil, err
 	}
-	seedSnaps := seedMeta.Snaps
 
-	required := getAllRequiredSnapsForModel(model)
-	seeding := make(map[string]*seed.Snap, len(seedSnaps))
-	for _, sn := range seedSnaps {
-		seeding[sn.Name] = sn
+	essentialSeedSnaps := deviceSeed.EssentialSnaps()
+	seedSnaps, err := deviceSeed.ModeSnaps("run") // XXX mode should be passed in
```

Comment:
Is the comment relevant? You're passing mode, no?

Reply:
it's about passing mode into populate* itself